Set up puppeteer tests with Firefox.

Riken-Shah opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
#16240


**Testing plan:** <!-- How have you tested? -->
Tested locally.


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
Here is the screenshot of "Not Secure Warning" hiding the Get API Key button.
Refer this [commit](https://github.com/zulip/zulip/pull/17136/commits/428f55bb3964be9dcea4a91dbf46b629a8f500ce)
![Screenshot 2021-02-17 at 12 25 13 AM](https://user-images.githubusercontent.com/63820270/108224108-8e1d0e00-7160-11eb-9ce8-b6229992398e.png)



<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on frontend_tests/puppeteer_lib/common.js:
```diff
@@ -66,7 +66,8 @@ class CommonUtils {
                     "--disable-setuid-sandbox",
                 ],
-                defaultViewport: {width: 1280, height: 1024},
+                // TODO: Change defaultViewport to 1280x1024 when puppeteer fixes the window size issue with firefox.
```
Can you link to an issue tracking the Puppeteer/firefox issue?

Riken-Shah replied:
@timabbott I have added the link to the issue.

timabbott commented:
@chdinesh1089 @priyank-p can you do a review on this?

Riken-Shah commented:
I have rebased the PR!
cc: @chdinesh1089 @priyank-p

chdinesh1089 commented on frontend_tests/puppeteer_lib/common.js:
```diff
@@ -435,7 +435,10 @@ class CommonUtils {
-        const url = page.url();
+        // `page.url()` does not show url fragment when running with Firefox.
+        // Here is the link to the issue https://github.com/puppeteer/puppeteer/issues/6787.
+        // TODO: When the above issue is solved please replaced the below line with `page.url()`.
```
typo here and in other copies of this comment: replace

chdinesh1089 commented on frontend_tests/puppeteer_tests/16-settings.js:
```diff
@@ -80,6 +80,8 @@ async function test_get_api_key(page) {
     await common.fill_form(page, "#api_key_form", {
         password: test_credentials.default_user.password,
     });
+    // To avoid flake sometimes when running with Firefox.
+    await page.focus(get_api_key_button_selector);
     await page.click(get_api_key_button_selector);
```
I think it'd be better to mention what causes the flake and how this fixes it in the commit message.

chdinesh1089 commented:
Looks good to me. Added two comments which wouldn't require any code changes.

Riken-Shah commented:
@chdinesh1089 Updated the PR!

timabbott commented on frontend_tests/puppeteer_lib/common.js:
```diff
@@ -435,7 +435,10 @@ class CommonUtils {
+        // `page.url()` does not show url fragment when running with Firefox.
+        // Here is the link to the issue https://github.com/puppeteer/puppeteer/issues/6787.
+        // TODO: When the above issue is solved please replace the below line with `page.url()`.
+        const url = await page.evaluate("location.href");
```
Can we extract this to a `common.page_url()` method we call from 3 places?  Would save on copy/pasting the comment.

Riken-Shah replied:
@timabbott I suggest we name the function `common.page_url_with_fragment()` because `page.url()` works fine when we don't need URL fragment.

Riken-Shah replied:
@timabbott I suggest we name the function `common.page_url_with_fragment()` because `page.url()` works fine when we don't need URL fragment.

timabbott replied:
Yeah, that works.

timabbott commented:
Posted one comment; otherwise this lgtm.  I tried running it locally and got a hang, so we can open a follow-up issue to investigate that, as long as it still works in Chrome.  The eventual goal is to be able to run it in CI; before we do that, we'll want to do the trick of running it in Firefox mode 100x in CI to see if it fails nondeterminstically?  That in turn can determine whether we want to enable running this mode in CI.  

@priyank-p @chdinesh1089 can help with how to do that?  I'm not sure if it's in our "working on puppeteer tests" documentation, but if not we should add it.

Riken-Shah commented:
@timabbott I have rebased the PR and updated the changes required.


> Posted one comment; otherwise this lgtm. I tried running it locally and got a hang, so we can open a follow-up issue to investigate that, as long as it still works in Chrome. The eventual goal is to be able to run it in CI; before we do that, we'll want to do the trick of running it in Firefox mode 100x in CI to see if it fails nondeterminstically? That in turn can determine whether we want to enable running this mode in CI.

Should I create one issue that will combine/point to all the issues or should I create separate issues?
BTW Where did it fail when you ran it?

timabbott commented:
I'd do a single master issue.

Can you rebase this PR for the merge conflicts?

Riken-Shah commented:
Rebased the PR @timabbott!

timabbott commented:
I merged about half the commits with some textual tweaks as the series ending with d12257e2c1a76877dd1ae33b18d528e8fdfb456d, and pushed the rest back here.  It's worth your reading the commit messages and comments I adjusted, as it may help you improve your technical writing style.

Riken-Shah commented:
> I merged about half the commits with some textual tweaks as the series ending with [d12257e](https://github.com/zulip/zulip/commit/d12257e2c1a76877dd1ae33b18d528e8fdfb456d), and pushed the rest back here. It's worth your reading the commit messages and comments I adjusted, as it may help you improve your technical writing style.

Ok, Tim. I'll work on it and try to improve it.

Riken-Shah commented:
I think I know why it failed. 
This is because the overlay was not completely closed. 
I knew that because I ran the `07-navigation.ts` 100 times in Firefox mode. 
This fixes it [5d10e6fa](https://github.com/Riken-Shah/zulip/commit/5d10e6fadda93e5cbdec22df3aab0d180c8677e2) (Please avoid the commit message). 
It ran successfully for 100 times in [CircleCI](https://app.circleci.com/pipelines/github/Riken-Shah/zulip/172/workflows/c7074c35-d974-4381-be57-04e6bc600bf7/jobs/711) and [Github Actions](https://github.com/Riken-Shah/zulip/runs/2030592445).

Should I also add this commit?

timabbott commented:
I didn't merge the rest mostly because I only had a moment and wanted to have a bit more time to think about some of the others.

@Riken-Shah yeah, definitely add additional bug fix commits to this PR.

timabbott commented on frontend_tests/puppeteer_tests/02-message-basics.ts:
```diff
@@ -351,9 +349,9 @@ async function test_stream_search_filters_stream_list(page: Page): Promise<void>
-    );
+    await page.$eval(".stream-list-filter", (el) => {
+        (el as HTMLInputElement).value = "ver";
+    });
```
@Riken-Shah I think this one has a semantics change -- we're no longer focusing and typing in the box; I could imagine that affecting the behavior of the production code.  Though maybe we should just replace this with the standard Puppeteer functions for typing these characters into an input box, which will likely be better than either version here.

timabbott commented on frontend_tests/puppeteer_tests/06-edit.ts:
```diff
@@ -14,8 +14,12 @@ async function trigger_edit_last_message(page: Page): Promise<void> {
+    });
+    await page.$eval(".message_edit_content", (el) => {
+        (el as HTMLTextAreaElement).value = "";
+    });
```
Same with these -- should this just be some sort of `page.type` type command to clear the text boxes?

Riken-Shah replied:
I couldn't find something similar like that. 
I found this related [issue](https://github.com/puppeteer/puppeteer/issues/3094) but it was rejected by one of a maintainer of the project [here](https://github.com/puppeteer/puppeteer/issues/3094#issuecomment-581722198).

One alternative to this will be, 
```javascript
// This will select all the text.
await page.click("selector", {clickCount: 3});
await page.type("selector", "new text");
```

WDYT?

timabbott replied:
Yeah, that sounds better to me.  In general we want to do as little `page.$eval` as we can that could be better expressed via the builtins.

timabbott commented on frontend_tests/puppeteer_tests/16-settings.ts:
```diff
@@ -53,7 +53,11 @@ async function test_change_password(page: Page): Promise<void> {
+    // The below line is an alternative to that.
+    // TODO: Replace the below line with `await page.waitForSelector("#change_password_modal:focus", {visible: true})`
+    // when the above issue is resolved.
+    await page.waitForFunction(() => document.activeElement!.id === "change_password_modal");
```
Is there a `waitForFocus` type function we could use instead?

Riken-Shah replied:
No, I didn't find anything similar to `waitForFocus`.

timabbott replied:
OK, well, we can go with this if there isn't a better way to do it.

timabbott commented:
OK posted a batch of comments on the rest of this PR.  I'm excited about this!

Riken-Shah commented on frontend_tests/puppeteer_tests/07-navigation.ts:
```diff
@@ -34,6 +34,12 @@ async function navigate_to_settings(page: Page): Promise<void> {
 
     await page.click("#settings_page .content-wrapper .exit");
+    // Waits until the overlay is completely closed.
+    await page.waitForFunction(
```
I've used `page.waitForFunction` instead of `page.waitForSelector("selector", {visible: false})` because when I ran the puppeteer tests 100x, it seems to not wait until the overlay is completely closed.

Riken-Shah commented on frontend_tests/puppeteer_tests/07-navigation.ts:
```diff
@@ -34,6 +34,12 @@ async function navigate_to_settings(page: Page): Promise<void> {
 
     await page.click("#settings_page .content-wrapper .exit");
+    // Wait until the overlay is completely closed.
+    await page.waitForFunction(
```
I've used `page.waitForFunction` instead of `page.waitForSelector("selector", {visible: false})` because when I ran the puppeteer tests 100x, it seems not to wait until the overlay is completely closed.

andersk replied:
The right replacement is `waitForSelector` with `{hidden: true}`, not `{visible: false}`.

Riken-Shah commented:
@timabbott  a gentle reminder - This PR is ready for another review.

timabbott commented on frontend_tests/puppeteer_tests/03-compose.ts:
```diff
@@ -59,9 +59,9 @@ async function test_keyboard_shortcuts(page: Page): Promise<void> {
-        await page.$x(get_message_xpath("Compose stream reply test")),
-    );
+    const stream_message_xpath = get_message_xpath("Compose stream reply test");
+    await page.waitForXPath(stream_message_xpath, {visible: true});
```
@andersk should this one by `hidden: false` instead?

andersk replied:
No, both `visible: false` and `hidden: false` are default, so setting either to `false` again would never be useful. If you want one of them you have to set one of them to `true`.

(I haven’t checked whether we need the option here at all, though.)

Riken-Shah replied:
@andersk This is just a precaution from a rare flake I noticed [here](https://app.circleci.com/pipelines/github/Riken-Shah/zulip/145/workflows/b9652d26-18a8-40ff-9d90-6929c54b46b0/jobs/654) when I ran `03-compose.ts` for 100x in Firefox mode. 
I thought it failed because compose box closing and stream message click happened at the almost same time. So after adding these changes I ran it again and it ran successfully ([link to the workflow](https://app.circleci.com/pipelines/github/Riken-Shah/zulip/159/workflows/b543ba62-b1d7-4218-908e-cc90ebe5cb70/jobs/686)).

timabbott commented on frontend_tests/puppeteer_lib/common.ts:
```diff
@@ -104,6 +104,13 @@ class CommonUtils {
+    // This function will clear the existing value of the element and
+    // replace it with the text.
+    async clear_and_type(page: Page, selector: string, text: string): Promise<void> {
+        await page.click(selector, {clickCount: 3});
```
What does `clickCount: 3` do?  Maybe worth adding a comment above this if the idea is it clicks 3 times to select any text currently in the text box.

timabbott commented:
Posted a couple small comments, otherwise lgtm!

Riken-Shah commented:
@timabbott I have updated a PR!

andersk commented:
Merged, thanks @Riken-Shah!
